docs/01-programs/02-candy-machine/04-inserting-items.md

```diff
@@ -2,15 +2,200 @@
+
+If you do not have URIs for your items, you’ll first need to upload their JSON metadata one by one. This can either be using an off-chain solution — such as AWS or your own server — or an on-chain solution — such as Arweave or IPFS.
+
+Fortunately, our SDKs can help you with that. They allow you to upload a JSON object and retrieve its URI.
+
+Additionally, tools like [Sugar](/developer-tools/sugar/guides/sugar-for-cmv3) make uploading JSON metadata a breeze by uploading in parallel, caching the process and retrying failed uploads.
+
+<Accordion>
+<AccordionItem title="JS SDK" open={true}>
+<div className="accordion-item-padding">
+
+To upload some JSON metadata using the JS SDK, you should first select your prefered storage provider. By default, Bundlr will be used to upload to Arweave. Here’s how you can choose another one.
+
+```tsx
+// NFT.Storage is an external plugin so you must first install it:
+// npm install @metaplex-foundation/js-plugin-nft-storage
+
+import { nftStorage } from "@metaplex-foundation/js-plugin-nft-storage";
+metaplex.use(nftStorage());
+```
+
+Once you have selected a storage provider, you may use the `uploadMetadata` operation of the NFT module to upload some JSON metadata.
+
+```tsx
+// Uploading some JSON metadata contains assets already uploaded.
+const { uri } = await metaplex.nfts().uploadMetadata({
+  name: "My NFT #1",
+  description: "My description",
+  image: "https://arweave.net/123",
+});
+```
+
+If the assets inside your JSON metadata also need to be uploaded to the same storage provider, you may give a [`MetaplexFile`](https://metaplex-foundation.github.io/js/types/js.MetaplexFile.html) instead of a URI and it will upload it for you. For instance, here’s how you would upload some JSON metadata using a file uploaded in the browser.
+
+```tsx
+// Uploading some JSON metadata and its assets (here, in the browser).
+const { uri } = await metaplex.nfts().uploadMetadata({
+  name: "My NFT #1",
+  description: "My description",
+  image: await toMetaplexFileFromBrowser(event.target.files[0]),
+});
+```
+
+API References: [Operation](https://metaplex-foundation.github.io/js/classes/js.NftClient.html#uploadMetadata), [Input](https://metaplex-foundation.github.io/js/types/js.UploadMetadataInput.html), [Output](https://metaplex-foundation.github.io/js/types/js.UploadMetadataOutput.html).
+
+</div>
+</AccordionItem>
+</Accordion>
+
+## Inserting Items
+
+Now that we have a name and URI for all of our items, all we need to do is insert them into our Candy Machine account.
+
+Note that, because transactions are limited to a certain size, we cannot insert thousands of items within the same transaction. The number of items you can insert per transaction will depend on the **Name Length** and **URI Length** attributes defined in your Config Line Settings. The shorter your names and URIs are, the more you’ll be able to fit into a transaction.
```

Comment:
Perhaps we need to make it more explicit the effect of `Name Length` and `URI Length` values to the length of the `Name` and `URI` – i.e., the length of `Name` and `URI` are constraint to the `Name Length` and `URI Length` values.

Reply:
Maybe this need to be on the `Inserting Items Using Prefixes` section.

Reply:
Good point, I'll add that info somewhere.